### PR TITLE
X.509: Fix bug in SAN parsing and enhance negative testing

### PR DESCRIPTION
**Pull request by AndrzejKurek**
This PR is a continuation of https://github.com/Mbed-TLS/mbedtls/pull/2839, rebased on development and with a fixed Changelog.
This PR fixes https://github.com/Mbed-TLS/mbedtls/issues/2838 and adds numerous negative CRT parsing tests for the SubjectAlternativeName extension.
Changes compared to original: 
- Merge conflicts resolved (rather trivial);
- `MBEDTLS_ERROR_ADD` used instead of just adding errors;
- Changelog file instead of entry;

Backport: https://github.com/Mbed-TLS/mbedtls/pull/7063

### CODE REVIEW

**Review by mprse — COMMENTED** [review]
Code changes look good.

Checked coverage of the modified functions:

- `x509_get_subject_alt_name`
image

- `x509_get_other_name`
image

It seems that at least invalid tag in `x509_get_other_name` can be covered.

Regarding negative test cases. There is no information how binary data is modified.
In the new PRs we chose to modify DER format cert/csr files using script in `Makefile` and using files as tests inputs. But I'm not sure if there is sense to use this approach here looking on other test cases.

**Comment by AndrzejKurek** [comment]
> But I'm not sure if there is sense to use this approach here looking on other test cases.

It would make sense, and I would prefer to have such approach! But this is a PR that I took over that already had these tests in, and re-creating them would probably the most-time consuming task in this whole PR. I'll add a test for `x509_get_other_name`.

**Comment by AndrzejKurek** [comment]
> I'll add a test for `x509_get_other_name`

Ah, but the missing coverage you mentioned is pre-existing in code that wasn't touched here.
Also, the check is duplicated in `mbedtls_x509_parse_subject_alt_name`, so there's no chance that codepath will be reached through this function.
I think that, since it would require running not `mbedtls_x509_parse_subject_alt_name`, but `x509_get_other_name` directly with raw data, it would be better to not add this coverage in this PR and if you really insist on covering this line, it should be handled in a separate issue.
Do you agree?

**Comment by mprse** [comment]
> Do you agree?

 It's now documented in PR, so I'm fine with not covering this gap here.

**Review by mprse — APPROVED** [review]
All issues are resolved.
Approved.

**Comment by AndrzejKurek** [comment]
Rebased to resolve conflicts. @davidhorstmann-arm - are you planning to review this PR? :)

**Comment by davidhorstmann-arm** [comment]
> Rebased to resolve conflicts. @davidhorstmann-arm - are you planning to review this PR? :)

Yes, apologies - I have one pending comment, I'll get to finishing my review shortly.

**Review by mprse — APPROVED** [review]
Rebase looks fine.

**Review by d3zd3z — APPROVED** [review]
The code changes look good. I'm a little concerned that the negative test cases are just the blobs of data, but I don't know if it worth trying to document/reproduce how they were generated.

**Review by davidhorstmann-arm — COMMENTED** [review]
Ah, @d3zd3z has beaten me to it. I have just one comment (preexisting), plus the unreadability of the testcases. I'll remove myself as reviewer since there are already 2.

File: library/x509.c [hunk 1]
```diff
@@ -1238,8 +1238,6 @@ int mbedtls_x509_get_subject_alt_name(unsigned char **p,
 {
     int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
     size_t len, tag_len;
-    mbedtls_asn1_buf *buf;
-    unsigned char tag;
     mbedtls_asn1_sequence *cur = subject_alt_name;
```
[the remark this answers is not in the document]
Comment (AndrzejKurek) [reply library/x509.c:1239]: I would be happy to introduce it in a separate PR, since it's pre-existing.

[the remark this answers is not in the document]
Comment (AndrzejKurek) [reply library/x509.c:1239]: It seems to me that both places that this is used - in the CSR and CRT instance, the passed `mbedtls_x509_sequence` is a part of certificate / CSR, which are memset to zero upon initialization. It's an internal function, so hopefully that's enough?

**Pull request merged**